### DISCUSSION TITLE
How to assign permissions to SSO users?

### DISCUSSION THREAD

**Discussion by yash-ahir** [inventree/InvenTree#7200] [Q&A] [answered]
Is there any way to import the groups supplied with SSO (I'm using `openid_connect` as the provider) or auto assign groups to SSO users? Right now when a new user signs in using SSO, the interface is completely blank as they don't have any permissions. I have groups `invetree_staff` and `invetree_admin`, I want to assign staff and admin permissions to users in those groups respectively.

**Comment by SchrodingersGat** [comment, marked as answer]
I do not believe this is currently possible. Refer to these docs:

- https://docs.inventree.org/en/latest/settings/SSO/
- https://docs.allauth.org/en/latest/socialaccount/providers/openid_connect.html

Extending this functionality would probably be a good target for a custom plugin - handling all possible edge cases here would be very complex.

**Reply by p-fruck** [reply]
@SchrodingersGat I am surprised you'd recommend developing a custom plugin here. I am running an InvenTree instance with similar requirements, currently we are assigning all groups manually which is only possible since the amount of users is rather low. I've been considering a similar feature, my idea would be to assign a group mapping dictionary (OIDC attribute to group name) and forcefully sync the groups on each login, i.e. all groups are set by the corresponding attribute and if no attribute matches the default group is assigned. I thought integrating such functionality into the InvenTree codebase would be the easiest approach?
Also, is the something like an AuthenticationEvent for the EventMixin or how would one hook into the authentication process using a plugin?

**Reply by p-fruck** [reply]
Update: Quickly hacked together an initial working draft. Instructions and code can be found here: https://github.com/p-fruck/inventree-sso-groups-plugin
Let me know what you think, I would love to see such functionality in InvenTree

**Reply by SchrodingersGat** [reply]
@p-fruck my suggestion to make this a plugin is because the required functionality may vary from one SSO provide to another. However looking at your suggested plugin, it looks to be quite generic? It may work as first-party supported feature in InvenTree after all

@matmair what are your thoughts on this one?

**Reply by p-fruck** [reply]
Yeah, imo the main challenge here is for the IdP and InvenTree admin(s) to do the correct configuration (claim mapping to roles depending on user store etc) whereas the implementation part for InvenTree can be held quite generic

**Reply by SchrodingersGat** [reply]
@p-fruck I'd happily support a PR to the core codebase if you want to submit one. It will require updates to the documentation also :)

**Reply by p-fruck** [reply]
Sounds great! Any idea where I should put the signals part thats in the __init__ method of the plugin? Should I keep relying on allauth signals or do you prefer using a custom hook in the codebase?

**Reply by SchrodingersGat** [reply]
I would suggest you add `events.py` to `src/backend/InvenTree/users/` - and look at how we structure `events.py` in the other code subdirectories.

**Reply by p-fruck** [reply]
I don't understand how I should use the events.py. It seems like this file only reexports methods to trigger/process custom events. What I am interested in are the two receivers at the end of the base/events/events.py and how to register them: https://github.com/inventree/InvenTree/blob/master/src/backend/InvenTree/plugin/base/event/events.py#L159-L186

From the [django docs](https://docs.djangoproject.com/en/5.0/topics/signals/#module-django.dispatch) I would connect the receivers in the `ready` method of `apps.py`, but I dont see that happening for your `plugin/events.py`

Also, do you want the code to be placed in the `users` app or in the `sso.py`?

**Reply by SchrodingersGat** [reply]
I forgot about `sso.py` - that's  a good location for the code.

So, in summary:

- Receivers are connected in `./src/backend/InvenTree/InvenTree/apps.py`
- Callbacks are added in `./src/backend/InvenTree/InvenTree/sso.py`
